### PR TITLE
Metricbeat Kubernetes module and metricsets docs proposal

### PR DESCRIPTION
**Pull request by eedugon**
As agreed with @exekias , I'm adding in this PR some changes in Kubernetes module main document and in the majority of metricsets, including descriptions and configuration details for each one.

### CODE REVIEW

**Comment by eedugon** [comment]
@exekias , please review this PR. I'm going to need some help with some links, and I'm not totally sure that I have modified the right file.
I'm doubting between the file I have modified (`metricbeat/docs/modules/kubernetes.asciidoc`) and (`metricbeat/module/kubernetes/_meta/docs.asciidoc`)

**Comment by exekias** [comment]
@ChrsMark could you have a look to this one and help @eedugon to move it forward :pray: ?

**Comment by ChrsMark** [comment]
Hey @eedugon! The change must be applied on `metricbeat/module/kubernetes/_meta/docs.asciidoc` and then run `mage update` under `metricbeat` directory so as to automatically update  `metricbeat/docs/modules/kubernetes.asciidoc`.

**Comment by eedugon** [comment]
@ChrsMark , @exekias , I think i have fixed the CLA issue and I have made the changes into the right file.
As you said `mage update` also changed the final file too. This latest commit includes the content after running the update too.
Let me know your thoughts.

This would require also updating 7.x docs, but maybe it's better to review them in master (`docs-preview` doesn't work for me), and as soon as all looks great we can do the backport.

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 1]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
 - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/[scheduler]
 - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/[proxy]
 
-All metricsets with the `state_` prefix require `hosts` field pointing to kube-state-metrics
-service within the cluster, while the rest should be pointed to kubelet service. Check the
-example configuration on how to do it.
+Some of the previous components are running in each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:17]:
```suggestion
Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 2]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
...
 
-The default metricsets are `container`, `node`, `pod`, `system` and `volume`.
+For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes,Running Metricbeat on Kubernetes>> document.
+
+[float]
+=== Kubernetes endpoints and metricsets
+
+Kubernetes module is a bit complex as its internal metricsets require access to a wide variety of endpoints.
+
+This section highlights and introduces some groups of metricsets with similar endpoint access needs. For more details on the metricsets see the ****configuration example link*** and ***metricsets reference link**** sections below.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:26]: configuration example link?
metricsets reference link?

Should those be replaced by sth else?

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 3]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
...
+
+==== node / system / pod / container / module
+
+The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:30]:
```suggestion
The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` on each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 4]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
...
+The proxy metricset requires access to the proxy endpoint in each of Kubernetes nodes, hence it's recommended to configure it as a part of a `Metricbeat DaemonSet`.
+
+==== scheduler and controllermanager
+
+These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only in the master node, and they are not exposed via a Service, but there are different strategies available for its configuration:
+
+- Create `Kubernetes Services` to make `kube-controller-manager` and `kube-scheduler` available and configure the metricsets to point to these services as part of a Metricbeat `Deployment`.
+- Use `Autodiscovery` functionality as part of a Metricbeat DaemonSet and include the metricsets in a conditional template applied for the specific pods.
+
+Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running in the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:56]:
```suggestion
Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running on the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```

File: metricbeat/module/kubernetes/_meta/docs.asciidoc [hunk 5]
```diff
@@ -7,11 +7,46 @@ This module fetches metrics from Kubernetes several components:
 - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/[scheduler]
 - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/[proxy]
 
-All metricsets with the `state_` prefix require `hosts` field pointing to kube-state-metrics
-service within the cluster, while the rest should be pointed to kubelet service. Check the
-example configuration on how to do it.
+Some of the previous components are running in each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/_meta/docs.asciidoc:10]:
```suggestion
Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
```

File: metricbeat/module/kubernetes/_meta/docs.asciidoc [hunk 6]
```diff
@@ -7,11 +7,46 @@ This module fetches metrics from Kubernetes several components:
...
+[float]
+=== Kubernetes endpoints and metricsets
+
+Kubernetes module is a bit complex as its internal metricsets require access to a wide variety of endpoints.
+
+This section highlights and introduces some groups of metricsets with similar endpoint access needs. For more details on the metricsets see the ****configuration example link*** and ***metricsets reference link**** sections below.
+
+==== node / system / pod / container / module
+
+The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running in the hosts.
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/_meta/docs.asciidoc:23]:
```suggestion
The default metricsets `container`, `node`, `pod`, `system` and `volume` require access to the `kubelet endpoint` in each of the Kubernetes nodes, hence it's recommended to include them as part of a `Metricbeat DaemonSet` or standalone Metricbeats running on the hosts.
```

File: metricbeat/module/kubernetes/_meta/docs.asciidoc [hunk 7]
```diff
@@ -7,11 +7,46 @@ This module fetches metrics from Kubernetes several components:
...
+The proxy metricset requires access to the proxy endpoint in each of Kubernetes nodes, hence it's recommended to configure it as a part of a `Metricbeat DaemonSet`.
+
+==== scheduler and controllermanager
+
+These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only in the master node, and they are not exposed via a Service, but there are different strategies available for its configuration:
+
+- Create `Kubernetes Services` to make `kube-controller-manager` and `kube-scheduler` available and configure the metricsets to point to these services as part of a Metricbeat `Deployment`.
+- Use `Autodiscovery` functionality as part of a Metricbeat DaemonSet and include the metricsets in a conditional template applied for the specific pods.
+
+Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running in the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/_meta/docs.asciidoc:49]:
```suggestion
Note: In some "As a Service" Kubernetes implementations, like `GKE`, the master nodes or even the pods running on the masters won't be visible. In these cases it won't be possible to use `scheduler` nor `controllermanager` metricsets.
```

**Comment by ChrsMark** [comment]
Left some minor comments. Content looks good overall! I also restarted the failed docs job.

File: metricbeat/docs/modules/kubernetes.asciidoc (hunk 2 above)
Comment (eedugon) [reply metricbeat/docs/modules/kubernetes.asciidoc:26]: @ChrsMark , I'd like to link there to the `Example Configuration` and `Metricsets` sections at the end of the final document. Take a look at the current https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-module-kubernetes.html

How should I add those links?

Comment (eedugon) [reply metricbeat/docs/modules/kubernetes.asciidoc]: Removing links and resolving conversation :)

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 8]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
 - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-scheduler/[scheduler]
 - https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/[proxy]
 
-All metricsets with the `state_` prefix require `hosts` field pointing to kube-state-metrics
-service within the cluster, while the rest should be pointed to kubelet service. Check the
-example configuration on how to do it.
+Some of the previous components are running on each of the Kubernetes nodes (like `kubelet` or `proxy`) while others provide a single cluster-wide endpoint. This is important to determine the optimal configuration and running strategy for the different metricsets included in the module.
 
-The default metricsets are `container`, `node`, `pod`, `system` and `volume`.
+For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes,Running Metricbeat on Kubernetes>> document.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:19]: nit 
```suggestion
For a complete reference on how to configure and run this module on Kubernetes as part of a `DaemonSet` and a `Deployment`, there's a complete example manifest available in <<running-on-kubernetes, Running Metricbeat on Kubernetes>> document.
```

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 9]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
...
+
+Depending on the version and configuration of Kubernetes nodes, `kubelet` might provide a read only http port (typically 10255), which is used in some configuration examples. But in general, and lately, this endpoint requires SSL (`https`) access (to port 10250 by default) and token based authentication.
+
+==== state_* and event
+
+All metricsets with the `state_` prefix require `hosts` field pointing to `kube-state-metrics
+service` within the cluster. As the service provides cluster-wide metrics, there's no need to fetch them per node, hence the recommendation is to run these metricsets as part of a `Metricbeat Deployment` with one only replica.
+
+Note: Kube-state-metrics is not deployed by default in Kubernetes. For these cases the instructions for its deployment are available [here](https://github.com/kubernetes/kube-state-metrics#kubernetes-deployment). Generally `kube-state-metrics` runs a `Deployment` and is accessible via a service called `kube-state-metrics` on `kube-system` namespace, which will be the service to use in our configuration.
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:39]: Urls are provided like this:
```suggestion
Note: Kube-state-metrics is not deployed by default in Kubernetes. For these cases the instructions for its deployment are available https://github.com/kubernetes/kube-state-metrics#kubernetes-deployment[here]. Generally `kube-state-metrics` runs a `Deployment` and is accessible via a service called `kube-state-metrics` on `kube-system` namespace, which will be the service to use in our configuration.
```

File: metricbeat/docs/modules/kubernetes.asciidoc [hunk 10]
```diff
@@ -14,11 +14,46 @@ This module fetches metrics from Kubernetes several components:
...
+
+The apiserver metricset requires access to the Kubernetes API, which should be easily available in all Kubernetes environments. Depending on the Kubernetes configuration, the API access might require SSL (`https`) and token based authentication.
+
+==== proxy
+
+The proxy metricset requires access to the proxy endpoint in each of Kubernetes nodes, hence it's recommended to configure it as a part of a `Metricbeat DaemonSet`.
+
+==== scheduler and controllermanager
+
+These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only in the master node, and they are not exposed via a Service, but there are different strategies available for its configuration:
```
Comment (ChrsMark) [inline metricbeat/docs/modules/kubernetes.asciidoc:51]:
```suggestion
These metricsets require access to the Kubernetes `controller-manager` and `scheduler` endpoints. By default, these pods run only on master nodes, and they are not exposed via a Service, but there are different strategies available for its configuration:
```

File: metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc [hunk 11]
```diff
@@ -1 +1,32 @@
...
+
+```yaml
+rules:
+- nonResourceURLs:
+  - /metrics
+  verbs:
+  - get
+```
+
+The previous configuration and RBAC requirement is available in the complete example manifest proposed in <<running-on-kubernetes,Running Metricbeat on Kubernetes>> document.
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc:32]:
```suggestion
The previous configuration and RBAC requirement is available in the complete example manifest proposed in <<running-on-kubernetes, Running Metricbeat on Kubernetes>> document.
```

File: metricbeat/docs/modules/kubernetes.asciidoc (hunk 10 above)
Comment (eedugon) [reply metricbeat/docs/modules/kubernetes.asciidoc]: Changed in my local branch and pushed

File: metricbeat/docs/modules/kubernetes.asciidoc (hunk 9 above)
Comment (eedugon) [reply metricbeat/docs/modules/kubernetes.asciidoc]: Changed in my local branch and pushed

File: metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc [hunk 12]
```diff
@@ -1 +1,32 @@
...
+This is the `apiserver` metricset of the Kubernetes module, in charge of retrieving metrics from the Kubernetes API (available at `/metrics`).
+
+This metricset needs access to the `apiserver` component of Kubernetes, accessible typically by any POD via the `kubernetes.default` service or via environment variables (`KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`).
+
+When the API uses https, the pod will need to authenticate using its default token and trust the server using the appropiate CA file.
+
+Configuration example using https and token based authentication:
+
+```yaml
+    - module: kubernetes
```
Comment (ChrsMark) [inline metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc:10]: Can you we fix the indentation here? It should be one shift left.

Comment (eedugon) [reply metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc]: pushed and solved

Comment (eedugon) [reply metricbeat/module/kubernetes/apiserver/_meta/docs.asciidoc]: ops... I forgot the docs update.... wait for it :)

**Pull request merged**